addons/stock/wizard/stock_quant_relocate.py

```diff
@@ -0,0 +1,76 @@
+    company_id = fields.Many2one(related="quant_ids.company_id")
+    location_id = fields.Many2one('stock.location', domain="[('usage', '=', 'internal'), ('company_id', '=', company_id)]")
+    allowed_package_ids = fields.Many2many('stock.quant.package', compute="_compute_allowed_package_ids")
+    package_id = fields.Many2one('stock.quant.package', domain="[('id', 'in', allowed_package_ids)]")
+    message = fields.Text('Reason for relocating stock')
+    is_not_full_package = fields.Boolean(compute='_compute_is_not_full_package')
+    is_not_single_location = fields.Boolean(compute='_compute_is_not_single_location')
+
+    def process(self):
+        if not self.location_id and not self.package_id:
+            return
+        self.quant_ids.action_set_inventory_quantity_to_zero()
+        # if quants have reserved quantities and are in a package, we have to unreserve the related stock.move.line
+        # because the package is removed before creating the move, which causes issues with reserved quantities in _free_reservation()
+        reserved_quants = self.quant_ids.filtered(lambda q: q.reserved_quantity > 0)
+        reserved_quants._unreserve_quants()
+
+        if self.location_id:
+            # remove the packages from the quant to avoid triggering a UserError in stock.move._action_done()
+            # we still keep them to reassign after the move
+            packs = self.quant_ids._get_package_dict()
+            # move the quants and get a dict with the old quant as key and the new quant as value
+            new_quants = self.quant_ids._move_quants(self.location_id, self.message or 'Product Relocated')
+
+            if self.package_id:
+                # assign the new package to the new quants
+                for new_quant in list(new_quants.values()):
+                    new_quant.package_id = self.package_id
+            else:
+                # reassign the old packages to the new quants
+                for quant in self.quant_ids:
+                    new_quants.get(quant.id).write({'package_id': list(filter(lambda k: quant.id in packs[k], packs))[0]})
+
+        elif self.package_id and not self.location_id:
+            # rewrite the current quants with the new package
+            self.quant_ids.package_id = self.package_id
+
+        if ("default_lot_id" in self.env.context):
+            return self.quant_ids.lot_id.action_lot_open_quants()
+        return self.env['ir.actions.server']._for_xml_id(self.env.context.get('action_ref', 'stock.action_view_quants'))
+
+    @api.depends('quant_ids')
+    def _compute_is_not_full_package(self):
+        self.is_not_full_package = False
+        packages = self.quant_ids.package_id
+        if packages and not all(qid in self.quant_ids.ids for qid in packages.quant_ids.ids):
+            self.is_not_full_package = True
+
+    @api.depends('location_id', 'quant_ids')
+    def _compute_is_not_single_location(self):
+        self.is_not_single_location = False
```

Comment:
It only make sense when a package is set

Reply:
no because the relocate wizard can also be used to simply assign a package without changing locations, so this compute is necessary to show the warning (or maybe just put the package_id field in readonly)